File: app/src/main/java/ep/rest/BookFormActivity.java
package ep.rest;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import java.io.IOException;

import okhttp3.Headers;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class BookFormActivity extends AppCompatActivity
        implements View.OnClickListener, Callback<Void> {
    private static final String TAG = BookFormActivity.class.getCanonicalName();

    private EditText author, title, price, description, year;
    private Button button;

    private Book book;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_book_form);

        author = (EditText) findViewById(R.id.etAuthor);
        title = (EditText) findViewById(R.id.etTitle);
        price = (EditText) findViewById(R.id.etPrice);
        year = (EditText) findViewById(R.id.etYear);
        description = (EditText) findViewById(R.id.etDescription);
        button = (Button) findViewById(R.id.button);
        button.setOnClickListener(this);

        final Intent intent = getIntent();
        book = (Book) intent.getSerializableExtra("ep.rest.book");
        if (book != null) {
            author.setText(book.author);
            title.setText(book.title);
            price.setText(String.valueOf(book.price));
            year.setText(String.valueOf(book.year));
            description.setText(book.description);
        }
    }

    @Override
    public void onClick(View view) {
        final String bookAuthor = author.getText().toString().trim();
        final String bookTitle = title.getText().toString().trim();
        final String bookDescription = description.getText().toString().trim();
        final double bookPrice = Double.parseDouble(price.getText().toString().trim());
        final int bookYear = Integer.parseInt(year.getText().toString().trim());

        if (book == null) { // dodajanje
            BookService.getInstance().insert(bookAuthor, bookTitle, bookPrice,
                    bookYear, bookDescription).enqueue(this);
        } else { // urejanje
            BookService.getInstance().update(book.id, bookAuthor, bookTitle, bookPrice,
                    bookYear, bookDescription).enqueue(this);
        }
    }

    @Override
    public void onResponse(Call<Void> call, Response<Void> response) {
        final Headers headers = response.headers();

        if (response.isSuccessful()) {
            final int id;
            if (book == null) {
                Log.i(TAG, "Insertion completed.");
                // Preberemo Location iz zaglavja
                final String[] parts = headers.get("Location").split("/");
                id = Integer.parseInt(parts[parts.length - 1]);
            } else {
                Log.i(TAG, "Editing saved.");
                id = book.id;
            }
            final Intent intent = new Intent(this, BookDetailActivity.class);
            intent.putExtra("ep.rest.id", id);
            startActivity(intent);
        } else {
            String errorMessage;
            try {
                errorMessage = "An error occurred: " + response.errorBody().string();
            } catch (IOException e) {
                errorMessage = "An error occurred: error while decoding the error message.";
            }
            Log.e(TAG, errorMessage);
        }
    }

    @Override
    public void onFailure(Call<Void> call, Throwable t) {
        Log.w(TAG, "Error: " + t.getMessage(), t);
    }
}

File: app/src/main/java/ep/rest/BookFormActivity.kt
package ep.rest

import android.content.Intent
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import ep.rest.databinding.ActivityBookFormBinding
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException

class BookFormActivity : AppCompatActivity(), Callback<Void> {

    private var book: Book? = null

    private val binding by lazy {
        ActivityBookFormBinding.inflate(layoutInflater)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(binding.root)

        binding.btnSave.setOnClickListener {
            val author = binding.etAuthor.text.toString().trim()
            val title = binding.etTitle.text.toString().trim()
            val description = binding.etDescription.text.toString().trim()
            val price = binding.etPrice.text.toString().trim().toDouble()
            val year = binding.etYear.text.toString().trim().toInt()

            if (book == null) { // dodajanje
                BookService.instance.insert(
                    author, title, price,
                    year, description
                ).enqueue(this)
            } else { // urejanje
                BookService.instance.update(
                    book!!.id, author, title, price,
                    year, description
                ).enqueue(this)
            }
        }

        val book = intent?.getSerializableExtra("ep.rest.book") as Book?
        if (book != null) {
            binding.etAuthor.setText(book.author)
            binding.etTitle.setText(book.title)
            binding.etPrice.setText(book.price.toString())
            binding.etYear.setText(book.year.toString())
            binding.etDescription.setText(book.description)
            this.book = book
        }
    }

    override fun onResponse(call: Call<Void>, response: Response<Void>) {
        val headers = response.headers()

        if (response.isSuccessful) {
            val id = if (book == null) {
                // Preberemo Location iz zaglavja
                Log.i(TAG, "Insertion completed.")
                val parts =
                    headers.get("Location")?.split("/".toRegex())?.dropLastWhile { it.isEmpty() }
                        ?.toTypedArray()
                // spremenljivka id dobi vrednost, ki jo vrne zadnji izraz v bloku
                parts?.get(parts.size - 1)?.toInt()
            } else {
                Log.i(TAG, "Editing saved.")
                // spremenljivka id dobi vrednost, ki jo vrne zadnji izraz v bloku
                book!!.id
            }

            val intent = Intent(this, BookDetailActivity::class.java)
            intent.putExtra("ep.rest.id", id)
            startActivity(intent)
        } else {
            val errorMessage = try {
                "An error occurred: ${response.errorBody()?.string()}"
            } catch (e: IOException) {
                "An error occurred: error while decoding the error message."
            }

            Log.e(TAG, errorMessage)
        }
    }

    override fun onFailure(call: Call<Void>, t: Throwable) {
        Log.w(TAG, "Error: ${t.message}", t)
    }

    companion object {
        private val TAG = BookFormActivity::class.java.canonicalName
    }
}
